libs/ml/include/ml/ops/ops_lookup.hpp

```diff
@@ -0,0 +1,254 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/graph.hpp"
+#include "ml/ops/activations/dropout.hpp"
+#include "ml/ops/activations/elu.hpp"
+#include "ml/ops/activations/leaky_relu.hpp"
+#include "ml/ops/activations/logsigmoid.hpp"
+#include "ml/ops/activations/logsoftmax.hpp"
+#include "ml/ops/activations/randomized_relu.hpp"
+#include "ml/ops/activations/relu.hpp"
+#include "ml/ops/activations/sigmoid.hpp"
+#include "ml/ops/activations/softmax.hpp"
+#include "ml/ops/add.hpp"
+#include "ml/ops/concatenate.hpp"
+#include "ml/ops/convolution_1d.hpp"
+#include "ml/ops/convolution_2d.hpp"
+#include "ml/ops/divide.hpp"
+#include "ml/ops/embeddings.hpp"
+#include "ml/ops/exp.hpp"
+#include "ml/ops/flatten.hpp"
+#include "ml/ops/log.hpp"
+#include "ml/ops/loss_functions.hpp"
+#include "ml/ops/matrix_multiply.hpp"
+#include "ml/ops/max_pool_1d.hpp"
+#include "ml/ops/max_pool_2d.hpp"
+#include "ml/ops/maximum.hpp"
+#include "ml/ops/multiply.hpp"
+#include "ml/ops/placeholder.hpp"
+#include "ml/ops/reshape.hpp"
+#include "ml/ops/sqrt.hpp"
+#include "ml/ops/subtract.hpp"
+#include "ml/ops/tanh.hpp"
+#include "ml/ops/transpose.hpp"
+#include "ml/ops/weights.hpp"
+
+#include <typeindex>
+
+namespace fetch {
+namespace ml {
+namespace ops {
+
+template <typename GraphType, typename Op>
+void MyAddNode(GraphType *g_ptr,  // nb: needs to be raw pointer because it's called with "this"
+               std::shared_ptr<SaveableParams> saved_node, std::string node_name,
+               std::vector<std::string> const &inputs)
+{
+  auto castnode = std::dynamic_pointer_cast<typename Op::SPType>(saved_node);
+  if (!castnode)
+  {
+    throw std::runtime_error("Node downcasting failed.");
```

Comment:
are we sure it's not possible to do this with static casts instead of dynamic casts?
then we can get rid of the throw and have an assert